### DISCUSSION TITLE
Binary threshold on saturation instead of luminescence value

### DISCUSSION THREAD

**Discussion by Franz1960** [SixLabors/ImageSharp#1539] [Ideas] [answered]
In a biological monitoring project we analyse belly patterns of yellow bellied toads (bombina variegata). We need to find patterns of yellow and black colors. Unfortunately the black parts can range from very dark to light grey with luminescence values higher than the yellow parts. Here you can see the raw image of a toad:

![image](https://user-images.githubusercontent.com/19282511/107182062-584b8d00-69dc-11eb-8f0e-0edb37e2dcc4.png)

We use a rectangular field between cloak and snout, shown in the left part of the following image:

![image](https://user-images.githubusercontent.com/19282511/107182695-841b4280-69dd-11eb-9ead-c99424653b33.png)

The center image shows the BinaryThreshold processor applied with a threshold value of 0.5. I modified the processor to use the saturation value instead of the luminescence value for comparison with the threshold. This gives the right image, which discriminates between the yellow and the "black" parts.

Extending the processor and the corresponding constructors and extension methods I get the wanted results. The patch containing my extensions is here:

[ImageSharp_saturation_threshold.txt](https://github.com/SixLabors/ImageSharp/files/5941658/ImageSharp_saturation_threshold.txt)

If you think this solution makes sense for ImageSharp, I would be happy to find it integrated sometime.

BTW: Congratulations to this extremely elegant image processing library!

Cheers Franz

**Comment by brianpopow** [comment, marked as answer]
>  I modified the processor to use the saturation value instead of the luminescence value for comparison with the threshold. This gives the right image, which discriminates between the yellow and the "black" parts.

> If you think this solution makes sense for ImageSharp, I would be happy to find it integrated sometime.

This sound like a good addition to me. Would you mind creating a PR for that?

**Reply by JimBobSquarePants** [reply]
Agreed. We can augment the existing functionality in a non-breaking way. 

We need to find a fastest possible way of calculating the saturation quotient though as `ToHsl` might be a bit heavy going.

**Comment by Franz1960** [comment]
The HSL saturation component failing the purpose at the dark end of color space I examined other color features as scalar values to compare to the threshold. This value has to represent colorfulness or chromaticity to the human eye for my purposes. Using the YCbCr color model the distance of Cb or Cr to the achromatic level AL would be a candidate.

    MaxChroma = Max(Abs(Cb - AL), Abs(Cr - AL))

I created a test image with four different colors and gradients to black and white:
![image](https://user-images.githubusercontent.com/19282511/107849787-dbfee280-6dfd-11eb-83a0-cfe2bdccf9b5.png)
An ideal processor should classify the colorful central regions above threshold, the dark, gray and white regions below. The saturation component delivers this result for threshold = 0.25:
![image](https://user-images.githubusercontent.com/19282511/107850013-9a6f3700-6dff-11eb-9002-d34d8ba9b0da.png)
The light and colorful parts are white (above threshold), but so are the dark parts which have high saturation values but seem simply dark to the human eye.

The MaxChroma component defined earlier delivers this result:
![image](https://user-images.githubusercontent.com/19282511/107850093-284b2200-6e00-11eb-951f-625c63534d4d.png)

The regions that appear colored to the human eye are white here. It is exactly what we need to detect the belly pattern of our toads.

A branch containing the implementation and tests for color components Saturation and MaxChroma is ready. Should I make a PR?

**Reply by JimBobSquarePants** [reply]
Yep, please do so. We can work together to get it polished.

**Reply by antonfirsov** [reply]
I wonder what is the output for the dafault Luminance mode with the 25% threshold?

If users of BinaryTreshold favor visual accuracy over speed more likely, we can even consider changing the default mode.

**Reply by Franz1960** [reply]
Luminance mode with threshold 0.25 delivers
![image](https://user-images.githubusercontent.com/19282511/107851905-8f230800-6e0d-11eb-9a9a-804f3d8a7136.png)
I would not change default mode, because I would expect from a binary threshold discriminator to discriminate on lightness.

**Reply by antonfirsov** [reply]
Thinking it through again, actually, depends on use-case doesn't it?

Other users might need to extract the bright regions not the ones that are more colorful.
